DOC: Add notebook infrastructure for the docs

melissawm opened the pull request:
This is a first proposal for adding [MyST-NB notebooks](https://myst-nb.readthedocs.io/en/latest/) to the SciPy docs. 

This PR is still in draft as there are a few things still to be decided:
- If we need to run tests, we might need to look into [testbook](https://testbook.readthedocs.io/en/latest/index.html) instead of [nbval](https://nbval.readthedocs.io/en/latest/) (see https://github.com/executablebooks/MyST-NB/issues/290 for details)
- I did some editing on the front page of the user guide. A lot of the contents are redundant now that the documentation has grown, and I felt like we could simplify them to read more like technical documentation and less like a book. I'm open to feedback or to breaking this into a separate PR if that's easier to review.
- We could also add a guide on opening the tutorial as a Jupyter notebook.
- Still pending is adding examples of cross-referencing the rst documentation and the md file.
- The notebook itself can still use some editing, but I just wanted to try a first build so that we can trace a plan.

#### Reference issue
See #5233

#### What does this implement/fix?
Adds myst_nb as a Sphinx extension, adds a new MyST notebook tutorial (as a markdown file) and also reorganizes the user guide landing page.

ev-br commented on doc/source/notebooks/interp_transition_guide.md:
````diff
@@ -0,0 +1,249 @@
+
+We start from the (slightly modified) docstring example.
+
+```{code-cell} ipython3
````
Is there a way to have this as a default, to avoid typing this for every code cell?

melissawm replied:
Clarified this with the notebook/myst-md conversion guide.

ev-br commented on doc/source/notebooks/interp_transition_guide.md:
````diff
@@ -0,0 +1,249 @@
+z.size == len(x) * len(y)
+```
+
+Also, note that `x.size != y.size`:
````
In ReST, there is a difference between double backticks for the literal code blocks and single backticks for intersphinx links. Which one would this be and how to generate both of these?

melissawm replied:
I added a few examples in the notebook, in short it's a bit more explicit but I hope it works anyway.

ev-br commented on doc/source/tutorial/index.rst:
```diff
@@ -4,16 +4,63 @@
 
+SciPy is a collection of mathematical algorithms and convenience functions built
+on NumPy_ . It adds significant power to the interactive Python session by
+providing the user with high-level commands and classes for manipulating and
```
Not only to interactive sessions, I suppose. (Minor nit, feel free to ignore.)

ev-br commented on doc/source/tutorial/index.rst:
```diff
@@ -4,16 +4,63 @@
+subpackages are also made available in the `scipy` namespace to ease
+their use in interactive sessions and programs. In addition, many
+basic array functions from :mod:`numpy` are also available at the
+top-level of the :mod:`scipy` package.
```
Not anymore?

ev-br reviewed (commented):
This looks great! 

> We could also add a guide on opening the tutorial as a Jupyter notebook.

Yes please! For a novice user, I've no idea how to start a tutorial. Do I type the markdown in a text editor (and then what), do I write in a jupyter notebook (and then what), how to pair the NB and md etc.

Also, would be great to have a guidance on how to make intersphinx links both from a new-style md document into existing rst pages and vice versa.

tupui reviewed (commented):
That's great! +1 for this and the editing you've done to existing content. Just some suggestions.

I would not spend too much time on the review of the tutorial itself and leave this for later. We can even say nobody looks at this, it's just to see how it renders and we remove it before merging (my preference). Then we can do a proper PR to add content and take our time.

melissawm commented on environment.yml:
```diff
@@ -38,6 +38,9 @@ dependencies:
   - sphinx-design
+  - docutils<0.18.1
+  #- jupytext
+  - myst-nb
```
I think I got it all

melissawm commented on doc/source/notebooks/interp_transition_guide.md:
```diff
@@ -0,0 +1,249 @@
+---
+jupytext:
+  text_representation:
+    extension: .md
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.14.0
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
```
Clarified this with the notebook/myst-md conversion guide.

melissawm commented:
I have [created an issue in MyST-NB about the `JUPYTER_PLATFORM_DIRS` warning](https://github.com/executablebooks/MyST-NB/issues/462). This is fixable if we set the environment variable as the warning describes, but I'd like to understand what's happening anyway.

tupui commented on doc/source/tutorial/index.rst:
```diff
@@ -28,6 +68,33 @@ SciPy User Guide
    ndimage
    io
 
+.. note::
+
+   In this documentation, we will often assume that the main packages (NumPy,
+   SciPy, and Matplotlib) have been imported as::
+
+       >>> import numpy as np
+       >>> import matplotlib as mpl
+       >>> import matplotlib.pyplot as plt
+
+   While we obviously don't require you to follow these conventions in your own
+   code, they are highly recommended.
```
I would maybe remove this paragraph since we want all code samples to be self sufficient now.

tupui commented on environment.yml:
```diff
@@ -38,6 +38,9 @@ dependencies:
   - matplotlib
   - pydata-sphinx-theme==0.9.0
   - sphinx-design
+  - docutils<0.18.1
+  #- jupytext
+  - myst-nb
```
Leftover?
```suggestion
  - docutils<0.18.1
  - myst-nb
```

tupui commented on environment.yml:
```diff
@@ -7,7 +7,7 @@ name: scipy-dev
   - conda-forge
 dependencies:
-  - python
+  - python<=3.9
```
I would not change this if it's only for the doc. This file is used for normal dev env creation.
```suggestion
  - python
```

tupui reviewed (commented):
Thanks for the update @melissawm! I have some minor suggestions but otherwise LGTM

melissawm commented on .gitignore:
```diff
@@ -129,6 +129,7 @@ Thumbs.db
 #################################
 doc/frontpage/build
 doc/source/reference/generated
+doc/source/notebooks/.ipynb_checkpoints
```
I am not sure which folders you mean?

tupui replied:
I mean changing this to
```suggestion
**/.ipynb_checkpoints
```

tupui commented:
I think failure in CI is related. Seems to be due to a deprecation warning in SQLAlchemy and raised by `myst`

melissawm commented:
It is 🥲 I also observed that locally. I will investigate

tupui commented:
If it's just warnings, we can selectively disable them in the config file for the doc (around L150).

melissawm commented:
Ok! Here's the update:
- I've merged main to make sure all was up to date and fix conflicts.
- I've updated the requirements and made sure the notebook runs smoothly. Here's the reason for the changes to conf.py and dev.py:

1. Several sqlalchemy errors appear, related to myst_nb and a few deprecations marked to happen in sqlalchemy 2.0. Pinning sqlalchemy makes no difference, so I've added the `SQLALCHEMY_SILENCE_UBER_WARNING=1` to the env variables for the doc task in dev.py
   <details>

	```
	python dev.py doc
	💻  ninja -C /home/melissa/projects/scipy/build
	ninja: Entering directory `/home/melissa/projects/scipy/build'
	[2/2] Generating scipy/generate-config with a custom command
	Build OK
	💻  meson install -C build --only-changed
	Installing, see meson-install.log...
	Installation OK
	# for testing
	# @echo installed scipy 591765c matches git version 591765c; exit 1
	mkdir -p build/html build/doctrees
	LANG=C /opt/miniconda/envs/scipy-dev/bin/python -msphinx -WT --keep-going  -b html -d build/doctrees -j1 source build/html 
	Running Sphinx v5.3.0
	SciPy (VERSION 1.10.0.dev0+2284.591765c)
	
	Traceback (most recent call last):
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/cmd/build.py", line 276, in build_main
	    app = Sphinx(args.sourcedir, args.confdir, args.outputdir,
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/application.py", line 223, in __init__
	    self.setup_extension(extension)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/application.py", line 398, in setup_extension
	    self.registry.load_extension(self, extname)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/registry.py", line 472, in load_extension
	    metadata = setup(app)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_nb/__init__.py", line 8, in setup
	    from .sphinx_ext import sphinx_setup
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_nb/sphinx_ext.py", line 20, in <module>
	    from myst_nb.ext.eval import load_eval_sphinx
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_nb/ext/eval/__init__.py", line 10, in <module>
	    from myst_nb.core.execute.base import EvalNameError
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_nb/core/execute/__init__.py", line 7, in <module>
	    from .cache import NotebookClientCache
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_nb/core/execute/cache.py", line 11, in <module>
	    from jupyter_cache.base import CacheBundleIn
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/jupyter_cache/base.py", line 16, in <module>
	    from jupyter_cache.cache.db import NbCacheRecord, NbProjectRecord
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/jupyter_cache/cache/db.py", line 17, in <module>
	    OrmBase = declarative_base()
	  File "<string>", line 2, in declarative_base
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sqlalchemy/util/deprecations.py", line 467, in warned
	    _warn_with_version(message, version, wtype, stacklevel=3)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sqlalchemy/util/deprecations.py", line 47, in _warn_with_version
	    _emit_uber_warning(type_, stacklevel)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sqlalchemy/util/deprecations.py", line 105, in _emit_uber_warning
	    _warnings_warn(warn, stacklevel=stacklevel + 1)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 1679, in _warnings_warn
	    warnings.warn(message, stacklevel=stacklevel + 1)
	sqlalchemy.exc.MovedIn20Warning: Deprecated API features detected! These feature(s) are not compatible with SQLAlchemy 2.0. To prevent incompatible upgrades prior to updating applications, ensure requirements files are pinned to "sqlalchemy<2.0". Set environment variable SQLALCHEMY_WARN_20=1 to show all deprecation warnings.  Set environment variable SQLALCHEMY_SILENCE_UBER_WARNING=1 to silence this message. (Background on SQLAlchemy 2.0 at: https://sqlalche.me/e/b8d9)
	
	Exception occurred:
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 1679, in _warnings_warn
	    warnings.warn(message, stacklevel=stacklevel + 1)
	sqlalchemy.exc.MovedIn20Warning: Deprecated API features detected! These feature(s) are not compatible with SQLAlchemy 2.0. To prevent incompatible upgrades prior to updating applications, ensure requirements files are pinned to "sqlalchemy<2.0". Set environment variable SQLALCHEMY_WARN_20=1 to show all deprecation warnings.  Set environment variable SQLALCHEMY_SILENCE_UBER_WARNING=1 to silence this message. (Background on SQLAlchemy 2.0 at: https://sqlalche.me/e/b8d9)
	The full traceback has been saved in /tmp/sphinx-err-rls3vz9x.log, if you want to report the issue to the developers.
	Please also report this if it was a user error, so that a better error message can be provided next time.
	A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
	make: *** [Makefile:110: html-build] Error 2
	```
   </details>
2. A Jupyter warning also needs the `JUPYTER_PLATFORM_DIRS=1` env variable set in dev.py.
	<details>
	
	```
	 python dev.py doc
	💻  ninja -C /home/melissa/projects/scipy/build
	ninja: Entering directory `/home/melissa/projects/scipy/build'
	[2/2] Generating scipy/generate-config with a custom command
	Build OK
	💻  meson install -C build --only-changed
	Installing, see meson-install.log...
	Installation OK
	# for testing
	# @echo installed scipy 591765c matches git version 591765c; exit 1
	mkdir -p build/html build/doctrees
	LANG=C /opt/miniconda/envs/scipy-dev/bin/python -msphinx -WT --keep-going  -b html -d build/doctrees -j1 source build/html 
	Running Sphinx v5.3.0
	SciPy (VERSION 1.10.0.dev0+2284.591765c)
	
	Traceback (most recent call last):
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/cmd/build.py", line 276, in build_main
	    app = Sphinx(args.sourcedir, args.confdir, args.outputdir,
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/application.py", line 223, in __init__
	    self.setup_extension(extension)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/application.py", line 398, in setup_extension
	    self.registry.load_extension(self, extname)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/registry.py", line 472, in load_extension
	    metadata = setup(app)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_nb/__init__.py", line 8, in setup
	    from .sphinx_ext import sphinx_setup
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_nb/sphinx_ext.py", line 20, in <module>
	    from myst_nb.ext.eval import load_eval_sphinx
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_nb/ext/eval/__init__.py", line 10, in <module>
	    from myst_nb.core.execute.base import EvalNameError
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_nb/core/execute/__init__.py", line 7, in <module>
	    from .cache import NotebookClientCache
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_nb/core/execute/cache.py", line 13, in <module>
	    from jupyter_cache.executors.utils import single_nb_execution
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/jupyter_cache/executors/utils.py", line 7, in <module>
	    from nbclient import execute as executenb
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/nbclient/__init__.py", line 6, in <module>
	    from .client import NotebookClient, execute  # noqa: F401
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/nbclient/client.py", line 14, in <module>
	    from jupyter_client import KernelManager
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/jupyter_client/__init__.py", line 8, in <module>
	    from .asynchronous import AsyncKernelClient  # noqa
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/jupyter_client/asynchronous/__init__.py", line 1, in <module>
	    from .client import AsyncKernelClient  # noqa
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/jupyter_client/asynchronous/client.py", line 9, in <module>
	    from ..client import KernelClient, reqrep
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/jupyter_client/client.py", line 20, in <module>
	    from .connect import ConnectionFileMixin
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/jupyter_client/connect.py", line 20, in <module>
	    from jupyter_core.paths import jupyter_data_dir, jupyter_runtime_dir, secure_write
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/jupyter_core/paths.py", line 208, in <module>
	    deprecation(
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/jupyter_core/utils/__init__.py", line 89, in deprecation
	    warnings.warn(message, DeprecationWarning, stacklevel=stacklevel + 1)
	DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
	given by the platformdirs library.  To remove this warning and
	see the appropriate new directories, set the environment variable
	`JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
	The use of platformdirs will be the default in `jupyter_core` v6
	
	Exception occurred:
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/jupyter_core/utils/__init__.py", line 89, in deprecation
	    warnings.warn(message, DeprecationWarning, stacklevel=stacklevel + 1)
	DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
	given by the platformdirs library.  To remove this warning and
	see the appropriate new directories, set the environment variable
	`JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
	The use of platformdirs will be the default in `jupyter_core` v6
	The full traceback has been saved in /tmp/sphinx-err-olo_4c9a.log, if you want to report the issue to the developers.
	Please also report this if it was a user error, so that a better error message can be provided next time.
	A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
	make: *** [Makefile:110: html-build] Error 2
	```
	</details>
3. Two docutils warnings are suppressed in `conf.py` 
	<details>
	
	```
	python dev.py doc      
	💻  ninja -C /home/melissa/projects/scipy/build
	ninja: Entering directory `/home/melissa/projects/scipy/build'
	[4/4] Generating scipy/generate-version with a custom command
	Build OK
	💻  meson install -C build --only-changed
	Installing, see meson-install.log...
	Installation OK
	# for testing
	# @echo installed scipy b36aa9f matches git version b36aa9f; exit 1
	mkdir -p build/html build/doctrees
	LANG=C /opt/miniconda/envs/scipy-dev/bin/python -msphinx -WT --keep-going  -b html -d build/doctrees -j1 source build/html 
	Running Sphinx v5.3.0
	SciPy (VERSION 1.11.0.dev0+1766.b36aa9f)
	[autosummary] generating autosummary for: dev/api-dev/api-dev-toc.rst, dev/api-dev/nan_policy.rst, dev/api-dev/special_ufuncs.rst, dev/conduct/code_of_conduct.rst, dev/conduct/report_handling_manual.rst, dev/contributor/adding_new.rst, dev/contributor/benchmarking.rst, dev/contributor/building.rst, dev/contributor/building_faq.rst, dev/contributor/compiled_code.rst, ..., tutorial/stats/discrete_zipf.rst, tutorial/stats/discrete_zipfian.rst, tutorial/stats/resampling.rst, tutorial/stats/sampling.rst, tutorial/stats/sampling_dau.rst, tutorial/stats/sampling_dgt.rst, tutorial/stats/sampling_hinv.rst, tutorial/stats/sampling_pinv.rst, tutorial/stats/sampling_srou.rst, tutorial/stats/sampling_tdr.rst
	[autosummary] generating autosummary for: /home/melissa/projects/scipy/doc/source/reference/generated/czt-function.rst, /home/melissa/projects/scipy/doc/source/reference/generated/odr-function.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.LowLevelCallable.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.cluster.hierarchy.ClusterNode.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.cluster.hierarchy.DisjointSet.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.cluster.hierarchy.average.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.cluster.hierarchy.centroid.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.cluster.hierarchy.complete.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.cluster.hierarchy.cophenet.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.cluster.hierarchy.correspond.rst, ..., /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.yeojohnson.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.yeojohnson_llf.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.yeojohnson_normmax.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.yeojohnson_normplot.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.yulesimon.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.zipf.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.zipfian.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.zmap.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.zscore.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.test.rst
	[autosummary] generating autosummary for: /home/melissa/projects/scipy/doc/source/reference/generated/scipy.LowLevelCallable.__getitem__.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.LowLevelCallable.__len__.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.LowLevelCallable.__mul__.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.LowLevelCallable.count.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.LowLevelCallable.from_cython.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.LowLevelCallable.function.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.LowLevelCallable.index.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.LowLevelCallable.signature.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.LowLevelCallable.user_data.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.cluster.hierarchy.ClusterNode.get_count.rst, ..., /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.sampling.NumericalInversePolynomial.set_random_state.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.sampling.NumericalInversePolynomial.u_error.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.sampling.SimpleRatioUniforms.rvs.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.sampling.SimpleRatioUniforms.set_random_state.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.sampling.TransformedDensityRejection.hat_area.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.sampling.TransformedDensityRejection.ppf_hat.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.sampling.TransformedDensityRejection.rvs.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.sampling.TransformedDensityRejection.set_random_state.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.sampling.TransformedDensityRejection.squeeze_area.rst, /home/melissa/projects/scipy/doc/source/reference/generated/scipy.stats.sampling.TransformedDensityRejection.squeeze_hat_ratio.rst
	loading intersphinx inventory from https://docs.python.org/3/objects.inv...
	loading intersphinx inventory from https://numpy.org/devdocs/objects.inv...
	loading intersphinx inventory from https://numpy.org/neps/objects.inv...
	loading intersphinx inventory from https://matplotlib.org/stable/objects.inv...
	loading intersphinx inventory from https://asv.readthedocs.io/en/stable/objects.inv...
	loading intersphinx inventory from https://www.statsmodels.org/stable/objects.inv...
	myst v0.18.1: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions=[], disable_syntax=[], all_links_external=False, url_schemes=('http', 'https', 'mailto', 'ftp'), ref_domains=None, highlight_code_blocks=True, number_code_blocks=[], title_to_header=False, heading_anchors=None, heading_slug_func=None, footnote_transition=True, words_per_minute=200, sub_delimiters=('{', '}'), linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area')
	myst-nb v0.17.1: NbParserConfig(custom_formats={}, metadata_key='mystnb', cell_metadata_key='mystnb', kernel_rgx_aliases={}, execution_mode='auto', execution_cache_path='', execution_excludepatterns=(), execution_timeout=30, execution_in_temp=False, execution_allow_errors=False, execution_raise_on_error=False, execution_show_tb=False, merge_streams=False, render_plugin='default', remove_code_source=False, remove_code_outputs=False, code_prompt_show='Show code cell {type}', code_prompt_hide='Hide code cell {type}', number_source_lines=False, output_stderr='show', render_text_lexer='myst-ansi', render_error_lexer='ipythontb', render_image_options={}, render_figure_options={}, render_markdown_format='commonmark', output_folder='build', append_css=True, metadata_to_fm=False)
	Using jupyter-cache at: /home/melissa/projects/scipy/doc/build/.jupyter_cache
	building [mo]: targets for 0 po files that are out of date
	building [html]: targets for 335 source files that are out of date
	updating environment: [new config] 4341 added, 0 changed, 0 removed
	reading sources... [  1%] notebooks/interp_transition_guide                                                                                  
	Traceback (most recent call last):
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/cmd/build.py", line 281, in build_main
	    app.build(args.force_all, args.filenames)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/application.py", line 347, in build
	    self.builder.build_update()
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 310, in build_update
	    self.build(to_build,
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 326, in build
	    updated_docnames = set(self.read())
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 433, in read
	    self._read_serial(docnames)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 454, in _read_serial
	    self.read_doc(docname)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 510, in read_doc
	    publisher.publish()
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/docutils/core.py", line 224, in publish
	    self.document = self.reader.read(self.source, self.parser,
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/sphinx/io.py", line 104, in read
	    self.parse()
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/docutils/readers/__init__.py", line 76, in parse
	    self.parser.parse(self.input, document)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_nb/sphinx_.py", line 142, in parse
	    mdit_renderer.setup_render(mdit_parser.options, mdit_env)
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_parser/mdit_to_docutils/base.py", line 135, in setup_render
	    self.document: nodes.document = options.get("document", make_document())
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/myst_parser/mdit_to_docutils/base.py", line 55, in make_document
	    settings = OptionParser(components=(parser_cls,)).get_default_values()
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/docutils/frontend.py", line 662, in __init__
	    warnings.warn('The frontend.OptionParser class will be replaced '
	DeprecationWarning: The frontend.OptionParser class will be replaced by a subclass of argparse.ArgumentParser in Docutils 0.21 or later.
	
	Exception occurred:
	  File "/opt/miniconda/envs/scipy-dev/lib/python3.10/site-packages/docutils/frontend.py", line 662, in __init__
	    warnings.warn('The frontend.OptionParser class will be replaced '
	DeprecationWarning: The frontend.OptionParser class will be replaced by a subclass of argparse.ArgumentParser in Docutils 0.21 or later.
	The full traceback has been saved in /tmp/sphinx-err-k3368iou.log, if you want to report the issue to the developers.
	Please also report this if it was a user error, so that a better error message can be provided next time.
	A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
	make: *** [Makefile:110: html-build] Error 2
	```
	</details>
4. I have also set up the notebook to ignore the `interp2d` deprecation warnings.

Let me know if this is acceptable - if it's too much extra stuff we can reevaluate, but most of these can and should be removed in the near future.

tupui commented on dev.py:
```diff
@@ -1035,6 +1035,11 @@ def task_meta(cls, list_targets, parallel, args, **kwargs):
         if parallel:
             make_params.append(f'SPHINXOPTS="-j{parallel}"')
 
+        # Environment variables needed for notebooks
+        # See gh-17322
+        make_params.append('SQLALCHEMY_SILENCE_UBER_WARNING=1')
+        make_params.append('JUPYTER_PLATFORM_DIRS=1')
```
I would be fine with these.

(In the same spirit of the dep tracker issue, maybe we should have a place to note things we need to do when something updates. Be it a Python version or another library. Otherwise this might stay there longer than intended, not that it would be a big issue though.)

tupui reviewed (commented):
Thanks for the update @melissawm!

I would be fine moving forward with this and silencing a few warnings seems ok to me.

There is still an issue with the CI. Did you see this locally as well?

melissawm commented:
I think this is a docutils issue - I don't see it locally but have found similar in other CIs. Let me check

rgommers commented:
@melissawm do you have a full traceback? If it's understood and it will be robust for Python >=3.10 we may consider just documenting the issue and moving on.

tupui reviewed (approved):
Awesome 👏 thanks for tracking this down! Approving on my side. We can try to discuss this tomorrow to get this in. Unless someone wants to merge now which is fine with me too OC 😃 


EDIT: folks and maintainers present at todays meeting were in favour of this change.

mdhaber commented:
This is awesome. Regarding questions 1-3 of [your email to the mailing list](https://mail.python.org/archives/list/scipy-dev@python.org/thread/X3MSMDTGMQDRGBN6TTN5CWG5H7QZG7O6/), I am not personally concerned with 1 (the new doc build dependencies), and I wouldn't want this PR to be held up by 2/3. Can those be addressed if we start to notice a problem?

How do you envision that new tutorials would be added? (e.g. would they all go in this executable notebook section at the bottom?  How would that be organized, and what would the relationship be with the existing tutorials?) 

Would it be worthwile to convert existing tutorials to this new format (maybe breaking them up into separate pages) so that new tutorials can be fit into the existing "User Guide"?

ev-br commented:
+1 from me. Would be great to do a test-drive while there's time until the next release.

Note to self to update the `interp2d` deprecation message to stop pointing to a random gist on somebody's GH account when this is merged.

rgommers commented:
This looks quite good to me - doc dependencies and build time don't seem to be problems I'd say. One question: there's mention of these as notebooks and being downloadable, but how? I don't see a link or button.

melissawm commented:
Hi all!

@mdhaber 
>  I am not personally concerned with 1 (the new doc build dependencies), and I wouldn't want this PR to be held up by 2/3. Can those be addressed if we start to notice a problem?

Absolutely, I'd be happy to revisit any of these decisions if we feel like performance is an issue or we want to move these out. Nothing here is irreversible.

> How do you envision that new tutorials would be added? (e.g. would they all go in this executable notebook section at the bottom? How would that be organized, and what would the relationship be with the existing tutorials?)

That is an initial proposal that I'm also happy to revisit later, I don't have strong opinions on this.

> Would it be worthwile to convert existing tutorials to this new format (maybe breaking them up into separate pages) so that new tutorials can be fit into the existing "User Guide"?

I'd say this is up for discussion. This is an easier way to share executable tutorials with users, but this also relates to the next point.

@rgommers 
> One question: there's mention of these as notebooks and being downloadable, but how? I don't see a link or button.

This is true - the PyData Sphinx theme, as I understand it, does not readily support the download buttons you would see on other themes (see, for example, https://numpy.org/numpy-tutorials/). However, each individual .md file here, if downloaded, can be opened by jupyter classic or jupyterlab as long as the jupytext extension is installed - so [downloading this file](https://raw.githubusercontent.com/melissawm/scipy/notebook-infra/doc/source/notebooks/interp_transition_guide.md) allows you to open it as a notebook and execute locally, edit it etc. This can absolutely be perfected here or in the future.

rgommers commented:
I'd just remove the mention on the front page of the tutorials then that these are downloadable notebooks - it's just going to lead to confusion. Once pydata-sphinx-theme gets support, the statement can be put back.

j-bowhay commented:
Also +1, looks like a nice addition

tupui commented:
Let's get this in 😃 thanks Melissa for the heavy lifting and everyone for the review and testing!

melissawm commented:
Thank you all!
